client/client.go

```diff
@@ -918,48 +1023,98 @@ func (r *NotaryRepository) bootstrapClient(checkInitialized bool) (*TUFClient, e
+// managing the key to the server. If key(s) are specified by keyList, then they are
+// used for signing the role.
 // These changes are staged in a changelist until publish is called.
-func (r *NotaryRepository) RotateKey(role string, serverManagesKey bool) error {
+func (r *NotaryRepository) RotateKey(role string, serverManagesKey bool, keyList []string) error {
 	if err := checkRotationInput(role, serverManagesKey); err != nil {
 		return err
 	}
-	var (
-		pubKey    data.PublicKey
-		err       error
-		errFmtMsg string
-	)
-	switch serverManagesKey {
-	case true:
+
+	pubKeyList, err := r.pubKeyListForRotation(role, serverManagesKey, keyList)
+	if err != nil {
+		return err
+	}
+
+	cl := changelist.NewMemChangelist()
+	if err := r.rootFileKeyChange(cl, role, changelist.ActionCreate, pubKeyList); err != nil {
+		return err
+	}
+	return r.publish(cl)
+}
+
+// Given a set of new keys to rotate to and a set of keys to drop, returns the list of current keys to use
+func (r *NotaryRepository) pubKeyListForRotation(role string, serverManaged bool, newKeys []string) (pubKeyList data.KeyList, err error) {
+	var pubKey data.PublicKey
+
+	// If server manages the key being rotated, request a rotation and return the new key
+	if serverManaged {
 		pubKey, err = rotateRemoteKey(r.baseURL, r.gun, role, r.roundTrip)
-		errFmtMsg = "unable to rotate remote key: %s"
-	default:
-		pubKey, err = r.CryptoService.Create(role, r.gun, data.ECDSAKey)
-		errFmtMsg = "unable to generate key: %s"
+		pubKeyList = make(data.KeyList, 0, 1)
+		pubKeyList = append(pubKeyList, pubKey)
+		if err != nil {
+			return nil, fmt.Errorf("unable to rotate remote key: %s", err)
+		}
+		return pubKeyList, nil
 	}
 
+	// If no new keys are passed in, we generate one
+	if len(newKeys) == 0 {
+		pubKeyList = make(data.KeyList, 0, 1)
+		pubKey, err = r.CryptoService.Create(role, r.gun, data.ECDSAKey)
+		pubKeyList = append(pubKeyList, pubKey)
+	}
 	if err != nil {
-		return fmt.Errorf(errFmtMsg, err)
+		return nil, fmt.Errorf("unable to generate key: %s", err)
 	}
 
-	// if this is a root role, generate a root cert for the public key
-	if role == data.CanonicalRootRole {
+	// If a list of keys to rotate to are provided, we add those
+	if len(newKeys) > 0 {
+		pubKeyList = make(data.KeyList, 0, len(newKeys))
+		for _, keyID := range newKeys {
+			pubKey = r.CryptoService.GetKey(keyID)
+			if pubKey == nil {
+				return nil, fmt.Errorf("unable to find key: %s")
+			}
+			pubKeyList = append(pubKeyList, pubKey)
+		}
+	}
+
+	// Convert to certs (for root keys)
+	if pubKeyList, err = r.pubKeysToCerts(role, pubKeyList); err != nil {
+		return nil, err
+	}
+
+	return pubKeyList, nil
+}
+
+func (r *NotaryRepository) pubKeysToCerts(role string, pubKeyList data.KeyList) (data.KeyList, error) {
+	// only generate certs for root keys
+	if role != data.CanonicalRootRole {
+		return pubKeyList, nil
+	}
+
+	for i, pubKey := range pubKeyList {
+		keyInfo, err := r.CryptoService.GetKeyInfo(pubKey.ID())
```

Comment:
Non-blocking optimization:  `r.CryptoService.GetPrivateKey(pubKey.ID())` returns the private key, the role, and an error, so we wouldn't need the extra call to `r.CryptoService.GetKeyInfo` to check that the role is root.